php/ext/google/protobuf/message.c

```diff
@@ -342,3 +343,322 @@ PHP_METHOD(Message, whichOneof) {
+
+    internal_add_generated_file(binary, binary_len, generated_pool TSRMLS_CC);
+    FREE(binary);
+  }
+
+  MessageHeader* intern = UNBOX(MessageHeader, getThis());
+  custom_data_init(any_type, intern PHP_PROTO_TSRMLS_CC);
+}
+
+PHP_METHOD(Any, getTypeUrl) {
+  zval member;
+  PHP_PROTO_ZVAL_STRING(&member, "type_url", 1);
+
+  zend_class_entry* old_scope = EG(scope);
+  EG(scope) = any_type;
+#if PHP_MAJOR_VERSION < 7
+  zval* value = message_handlers->read_property(getThis(), &member, BP_VAR_R,
+                                                NULL PHP_PROTO_TSRMLS_CC);
+#else
+  zval* value = message_handlers->read_property(getThis(), &member, BP_VAR_R,
+                                                NULL, NULL PHP_PROTO_TSRMLS_CC);
+#endif
+  EG(scope) = old_scope;
+
+#if PHP_MAJOR_VERSION < 7
+  ZVAL_ZVAL(return_value, value, 1, 0);
+#else
+  ZVAL_COPY(return_value, value);
+#endif
+}
+
+PHP_METHOD(Any, setTypeUrl) {
+  char *type_url = NULL;
+  PHP_PROTO_SIZE type_url_len;
+
+  if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "s", &type_url,
+                            &type_url_len) == FAILURE) {
+    return;
+  }
+
+  zval member;
+  zval value;
+  PHP_PROTO_ZVAL_STRING(&member, "type_url", 1);
+  PHP_PROTO_ZVAL_STRINGL(&value, type_url, type_url_len, 1);
+
+  zend_class_entry* old_scope = EG(scope);
+  EG(scope) = any_type;
+  message_handlers->write_property(getThis(), &member, &value,
+                                   NULL PHP_PROTO_TSRMLS_CC);
+  EG(scope) = old_scope;
+
+#if PHP_MAJOR_VERSION < 7
+  ZVAL_ZVAL(return_value, getThis(), 1, 0);
+#else
+  ZVAL_COPY(return_value, getThis());
+#endif
+}
+
+PHP_METHOD(Any, getValue) {
+  zval member;
+  PHP_PROTO_ZVAL_STRING(&member, "value", 1);
+
+  zend_class_entry* old_scope = EG(scope);
+  EG(scope) = any_type;
+#if PHP_MAJOR_VERSION < 7
+  zval* value = message_handlers->read_property(getThis(), &member, BP_VAR_R,
+                                                NULL PHP_PROTO_TSRMLS_CC);
+#else
+  zval* value = message_handlers->read_property(getThis(), &member, BP_VAR_R,
+                                                NULL, NULL PHP_PROTO_TSRMLS_CC);
+#endif
+  EG(scope) = old_scope;
+
+#if PHP_MAJOR_VERSION < 7
+  ZVAL_ZVAL(return_value, value, 1, 0);
+#else
+  ZVAL_COPY(return_value, value);
+#endif
+}
+
+PHP_METHOD(Any, setValue) {
+  char *value = NULL;
+  PHP_PROTO_SIZE value_len;
+
+  if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "s", &value,
+                            &value_len) == FAILURE) {
+    return;
+  }
+
+  zval member;
+  zval value_to_set;
+  PHP_PROTO_ZVAL_STRING(&member, "value", 1);
+  PHP_PROTO_ZVAL_STRINGL(&value_to_set, value, value_len, 1);
+
+  zend_class_entry* old_scope = EG(scope);
+  EG(scope) = any_type;
+  message_handlers->write_property(getThis(), &member, &value_to_set,
+                                   NULL PHP_PROTO_TSRMLS_CC);
+  EG(scope) = old_scope;
+
+#if PHP_MAJOR_VERSION < 7
```

Comment:
This pattern happens quite a lot. Perhaps worth it to put a macro in `protobuf.h`?